File: Problem_2_DP.java
// Time complexity - O()
// Space complexity - O()

// DP solution using dp array of length nums. Check for max values of nums[i]+i-2th index or nums[i-1].

class Solution {
    public int rob(int[] nums) {
    
    // Edge case
        if(nums==null||nums.length==0)
            return 0;
        if(nums.length == 1){
            return nums[0];
        }
        if(nums.length == 2){
            return Math.max(nums[0],nums[1]);
        }
        
        // Logic
        int[] dp = new int[nums.length];
        dp[0]=nums[0];
        dp[1]=Math.max(nums[0],nums[1]);
        for(int i=2; i<nums.length; i++){
            dp[i] = Math.max(nums[i]+dp[i-2], dp[i-1]);
        }
        return dp[nums.length-1];
    }
}

File: Problem_2_DP.cpp
//Space Complexity: O(NM), where n is number of houses and m is the number of choices we have.
//Time Complexity: O(n), where n is the length of the array.
//Executed successfully  on leetcode.

class Solution {
public:
    int rob(vector<int>& nums) {
        if(nums.size()==0) return 0;
        int dp[nums.size()][2];
        for(int i=0; i<nums.size(); i++){
            dp[i][0] = 0;
            dp[i][1] = 0;
        }
        dp[0][1] = nums.at(0);
        
        for(int i=1; i<nums.size(); i++){
            dp[i][0] = max(dp[i-1][0], dp[i-1][1]);
            dp[i][1] = dp[i-1][0] + nums.at(i);
        }
        int result = max(dp[nums.size()-1][0], dp[nums.size()-1][1]);
        return result;
    }
};